### ISSUE TITLE
dnsutil.go times-out with IPv6 with UDP, falling back to TCP, taking 20 seconds plus

### ISSUE THREAD

**Issue by cameronelliott** [caddyserver/certmagic#122]
TTTL;DR#3
I ran into an IPv6 Google Wifi resolver bug, it slows down the Lego/Certmagic DNS resolver (10 second queries),
but doesn't ultimately stop it.
My MacOs resolver was not confused or slowed down by the GoogleWifi issues.
It may not warrant re-doing, but if it did, a multi-mode Go net.Resolver which uses both the OS resolver and the native Go resolver would probably be more robust and faster. 
(In fact, one could resolve using 4 goroutines on these tuples: (os,ipv4) (os,ipv6) (go, ipv4) (go,ipv6)) to implement a fairly robust DNS resolver)

TL;DR#2
After discussing this with the [IPv6 group on Reddit](https://www.reddit.com/r/ipv6/comments/m6030s/dns_request_when_macos_googlewifi_has_ipv6/)
it really looks like I ran into a bug on the GoogleWifi which caused IPv6 responses to be returned from a different IP address that from which the request was received. On MacOs(well Linux too) the firewall won't let the packet back
in because the auto-pinhole UDP hole opener was expecting a response from a different IP address, so
the IPv6 answer never made it back to Certmagic.

A related issue here is that Certmagic and, well the original Lego code uses it's own hand crafted DNS resolver,
rather than supporting the operating-system resolver.
DNS resolution like Lego(Certmagic) does here is super-fiddly and has lots of land-mines.
The Golang DNS resolver in 'net' supports both the operating-system resolver and a pure-Go dns-resolver.
(Which can be switched at runtime via a flat in the net.Resolver struct)

I wrote a [Go DNS test program](https://github.com/cameronelliott/dns-issue) to see how Go vs. operating system does with this issue.

Here is the output:
```
duration: 1.68784ms name google.com num IPs found 2    usego:false network:ip
duration: 272.407µs name google.com num IPs found 1    usego:false network:ip4
duration: 219.011µs name google.com num IPs found 1    usego:false network:ip6
duration: 5.013298998s name google.com num IPs found 2    usego:true network:ip
duration: 5.016264918s name google.com num IPs found 1    usego:true network:ip4
duration: 5.016252006s name google.com num IPs found 1    usego:true network:ip6
```

Go also struggles with ipv6 on the broken Google Wifi, (5 second resolves!), but ultimately succeeds,
probably falling back to TCP like the Lego code does.

TL;DR:
dnsutil.go seems to work well with ipv4, I have noticed issues with using it with ipv6,
specifically, requests to the local dns server timeout and fail because of modified ipv6 source addresses
in responses to dnsutil.go DNS queries.
Dnsutil.go uses its own-handcrafted DNS query code rather than net.Resolver.LookupTXT() and net.Resolver.LookupIP, etc... which from my testing appear to be robust to these issues.

I am not sure why my local DNS server is changing it's response address when responding to IPv6 queries,
it is an up to date Google Wifi router. This behavior may be spec conformant DNS server behavior.

Actually, I suspect many users in ipv6 environments have 20 second registration times, because 
they have two ipv6 dns queries timeout with UDP, then fall back to TCP.

I don't expect you guys to fix this stuff, this is super fiddly-stuff, like playing drums with sticks of butter,
but there should be place to a discuss the issue if others run into issues with DNS01-ACME in ipv6 environments.

A few things to consider for the future:
- Use net.LookupIP("google.com") for NS/nameserver resolution, rather than hand-built DNS based on meikg/dns.
- Do concurrent ipv4 and ipv6 queries to find name servers.
- Do concurrent both meikg/dns queries with concurrent net.LookupIP to increase robustness.
- Do concurrent UDP and TCP queries.
- Consider re-writing this stuff.

---- long answer here ----

## What version of the package are you using?
go.mod:
github.com/caddyserver/certmagic v0.12.1-0.20210224184602-7550222c4a6a

## What are you trying to do?
I want to report issues discovered in dnsutil.go to help guide future changes
and understanding of pros and cons of dnsutil.go
Possibly also to create conversations about bugs and ways to fix.

## What steps did you take?
I am using functions from dnsutil.go in another project.
They generally work well under IPv4. 
it works great with IPv4, when I start using IPv6, it starts timing out.

## What did you expect to happen, and what actually happened instead?
I expected ipv6 usage not to wait 10 seconds for a timeout, then fallback to TCP.

## How do you think this should be fixed?
I honestly think all this code should be re-written to use net.Resolver.LookupXXX, and do concurrent
queries to be robust to differents between ipv4/ipv6, and manual-dns-queries vs net.resolver.lookupXXX()

## Please link to any related issues, pull requests, and/or discussion
https://www.reddit.com/r/ipv6/comments/m6030s/dns_request_when_macos_googlewifi_has_ipv6/

## Bonus: What do you use CertMagic for, and do you find it useful?
I love certmagic!

**Comment by cameronelliott** [comment]
**INFO UPDATE**

My testing and problems experienced were on a MacOs Big-Sur system.

The Go documentation accuses MacOs of restricting user programs from
being able to do their own DNS queries here:

https://golang.org/pkg/net/#hdr-Name_Resolution

> When cgo is available, the cgo-based resolver is used instead under a variety of conditions: on systems that do not let programs make direct DNS requests (OS X),...

My experience with Caddy and Certmagic do not support this assertion.
I mean does MacOs have special port-53 UDP firewall rules for inbound or outbound packets? I have seen zero evidence of this, and much evidence to the contrary. But it is worth noting.

**Comment by mholt** [comment]
Ok... wait, I'm lost.

What exactly is broken in Certmagic, and what needs to change?

**Comment by cameronelliott** [comment]
TL;DR
Nothing is broken in certmagic.
But certmagic could be more robust to external failures by concurrently trying both Ipv4 and Ipv6 resolver chains.

**Comment by mholt** [comment]
Ok cool, thanks for the simplification. :)

As long as we can use the OS resolver without calling out to C code, then :+1: -- feel free to submit a proposal in that regard.

> But certmagic could be more robust to external failures by concurrently trying both Ipv4 and Ipv6 resolver chains.

Would this prevent a failure or only improve performance? (Not that either is bad, just trying to understand.)
